crates/astria-core/src/protocol/transaction/v1alpha1/action_group.rs

```diff
@@ -0,0 +1,491 @@
+            ActionGroup::General(_) => write!(f, "General"),
+            ActionGroup::BundlableSudo(_) => write!(f, "BundlableSudo"),
+            ActionGroup::Sudo(_) => write!(f, "Sudo"),
+        }
+    }
+}
+
+impl From<BundlableGeneral> for ActionGroup {
+    fn from(val: BundlableGeneral) -> ActionGroup {
+        ActionGroup::BundlableGeneral(val)
+    }
+}
+
+impl From<General> for ActionGroup {
+    fn from(val: General) -> ActionGroup {
+        ActionGroup::General(val)
+    }
+}
+
+impl From<BundlableSudo> for ActionGroup {
+    fn from(val: BundlableSudo) -> ActionGroup {
+        ActionGroup::BundlableSudo(val)
+    }
+}
+
+impl From<Sudo> for ActionGroup {
+    fn from(val: Sudo) -> ActionGroup {
+        ActionGroup::Sudo(val)
+    }
+}
+
+#[derive(Debug, thiserror::Error)]
+enum ErrorKind {
+    #[error("input contains mixed action types")]
+    Mixed,
+    #[error("attempted to create bundle with non bundleable `ActionGroup` type")]
+    NotBundleable,
+}
+
+#[derive(Debug)]
+pub struct Error {
+    kind: ErrorKind,
+    context: Option<String>,
+}
+
+impl fmt::Display for Error {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{}", self.kind)?;
+        if let Some(ctx) = &self.context {
+            write!(f, ": {ctx}")?;
+        }
+        Ok(())
+    }
+}
+
+impl std::error::Error for Error {
+    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
+        Some(&self.kind)
+    }
+}
+
+impl Error {
+    fn new(kind: ErrorKind, context: Option<String>) -> Self {
+        Self {
+            kind,
+            context,
+        }
+    }
+
+    #[must_use]
+    pub fn mixed(
+        original_group: ActionGroup,
+        additional_group: ActionGroup,
+        action: &Action,
+    ) -> Self {
+        let context = format!(
+            "Mixed actions of different types. Original group: '{original_group}', Additional \
+             group: '{additional_group}', triggering action: '{}'",
+            action.name()
+        );
+        Self::new(ErrorKind::Mixed, Some(context))
+    }
+
+    #[must_use]
+    pub fn not_bundlable(group: ActionGroup) -> Self {
+        let context = format!("ActionGroup type '{group}' is not bundleable");
+        Self::new(ErrorKind::NotBundleable, Some(context))
+    }
+}
+
+/// Invariants: `group` is set if `inner` is not empty.
+#[derive(Clone, Debug)]
+pub(super) struct Actions {
+    inner: Vec<Action>,
+}
+
+impl Actions {
+    pub(super) fn actions(&self) -> &[Action] {
+        &self.inner
+    }
+
+    #[must_use]
+    pub(super) fn into_actions(self) -> Vec<Action> {
+        self.inner
+    }
+
+    pub(super) fn group(&self) -> Option<ActionGroup> {
+        self.inner.first().map(Group::group)
+    }
+
+    pub(super) fn default() -> Self {
+        Self {
+            inner: vec![],
+        }
+    }
+
+    pub(super) fn from_list_of_actions(actions: Vec<Action>) -> Result<Self, Error> {
```

Comment:
```suggestion
    pub(super) fn try_from_list_of_actions(actions: Vec<Action>) -> Result<Self, Error> {
```

since it returns a result